src/dbnode/storage/shard.go

```diff
@@ -2725,77 +2727,92 @@ func (s *dbShard) AggregateTiles(
 		return 0, err
 	}
 
 	var (
 		processedTileCount atomic.Int64
 		multiErr           xerrors.MultiError
-		dp                 ts.Datapoint
+		annotationPayload  annotation.Payload
+		downsampledValues  = make([]tile.DownsampledValue, 0, 4)
 		segmentCapacity    int
 		writerData         = make([][]byte, 2)
 	)
 
-READER:
 	for readerIter.Next() {
-		seriesIter, id, encodedTags := readerIter.Current()
+		var (
+			seriesIter, id, encodedTags = readerIter.Current()
+			prevFrameLastValue          = math.NaN()
+			hasData                     bool
+			handleValueResets           bool
+			firstAnnotation             ts.Annotation
+			writeErr                    error
+		)
 
 		for seriesIter.Next() {
 			frame := seriesIter.Current()
-			unit, singleUnit := frame.Units().SingleValue()
-			annotation, singleAnnotation := frame.Annotations().SingleValue()
 
 			if frameValues := frame.Values(); len(frameValues) > 0 {
-				lastIdx := len(frameValues) - 1
 
-				value := frameValues[lastIdx]
-				timestamp := frame.Timestamps()[lastIdx]
+				if !hasData {
+					annotationPayload.Reset()
+					firstAnnotation, err = frame.Annotations().Value(0)
+					if err != nil {
+						writeErr = err
+						break
+					}
 
-				if !singleUnit {
-					// TODO: what happens if unit has changed mid-tile?
-					// Write early and then do the remaining values separately?
-					if lastIdx < len(frame.Units().Values()) {
-						unit = frame.Units().Values()[lastIdx]
+					if annotationPayload.Unmarshal(firstAnnotation) == nil {
+						// Ignore the error if the annotation does not match the protobuf struct???
+						handleValueResets = annotationPayload.HandleValueResets
 					}
+					hasData = true
 				}
 
-				if !singleAnnotation {
-					// TODO: what happens if annotation has changed mid-tile?
-					// Write early and then do the remaining values separately?
-					if lastIdx < len(frame.Annotations().Values()) {
-						annotation = frame.Annotations().Values()[lastIdx]
+				downsampledValues = downsampledValues[:0]
+				lastIdx := len(frameValues) - 1
+
+				if handleValueResets {
+					// Last value plus possible few more datapoints to preserve counter semantics.
+					downsampledValues = tile.DownsampleCounterResets(prevFrameLastValue, frameValues, downsampledValues)
+				} else {
+					// Plain last value per frame.
+					downsampledValue := tile.DownsampledValue{
+						FrameIndex: lastIdx,
+						Value:      frameValues[lastIdx],
 					}
+					downsampledValues = append(downsampledValues, downsampledValue)
 				}
 
-				dp.Timestamp = timestamp
-				dp.Value = value
-				err = encoder.Encode(dp, unit, annotation)
-				if err != nil {
-					s.metrics.largeTilesWriteErrors.Inc(1)
-					multiErr = multiErr.Add(err)
-					break READER
+				writeErr = encodeDownsampledValues(downsampledValues, frame, firstAnnotation, encoder)
+				if writeErr != nil {
+					break
 				}
 
+				prevFrameLastValue = frameValues[lastIdx]
```

Comment:
nit: only do this `if handleValueResets` to make it clear this is only used in that logic?